feat: add support for selecting SSL key type (ECDSA/RSA)

mnr73 opened the pull request:
Added the ability to specify the SSL key type (ECDSA or RSA) for each site in the Nginx Proxy Manager. This enhancement is particularly useful for environments with IoT devices that have limitations with specific key types, such as RSA-only support. The implementation includes:

- Backend support for storing and validating the `ssl_key_type` field.
- Swagger schema updated to validate the new input.
- Frontend update to allow users to select the SSL key type via a dropdown menu.

This feature ensures greater flexibility and compatibility in managing SSL certificates for diverse setups.

mnr73 commented:
HI. Could you help me identify the issue so I can fix it?

Thanks!

mnr73 commented:
these errors are not related to my changes!

I think this is another problem. I build docker image with my change and test that. it's work trully.

mnr73 commented:
it's work truly :)
![image](https://github.com/user-attachments/assets/b9b5fe6a-92a6-4f64-a1c6-ce9bc51c52c0)
![image](https://github.com/user-attachments/assets/ca525670-24ad-4310-97fd-39c86fce6dd5)

mnr73 commented:
last cipher update result

![image](https://github.com/user-attachments/assets/6494d05a-8ab7-49d8-b913-4f547f14a22f)

jc21 commented:
Great work. Do you happen to have any automated commands that might test the feature? ie:
1. I request a ECDSA cert and run a console command to check it given I only have the HTTPS URL
2. I request a RSA cert and run a console command to check it given I only have the HTTPS URL

mnr73 commented:
Thanks. Yes I think could be test with nmap but I work on Diffie-Hellman cipher support and another problem that we had for IOT devices. after that I add some automated test for this.

Please hold on while I complete these.

mnr73 commented:
This pull request introduces features specifically designed to address common challenges with Embedded and IoT devices. The following updates have been made:

- Encryption Key Selection: Added the ability to choose between RSA and ECDSA encryption keys for each domain to enhance flexibility and security.
- Extended Cipher Suites: Included additional cipher suites to support a broader range of IoT devices.
- Support for Diffie-Hellman Parameters: Integrated cipher suites for stronger and more secure communications.
- Default Server Selection: Added the capability to designate a default server from existing hosts for simpler configuration and better management.

related issues
- #2539 
- #3354 
- #2147 
- #951

mnr73 commented:
the result supported ciphers

```
PORT    STATE SERVICE
443/tcp open  https
| ssl-enum-ciphers:
|   TLSv1.2:
|     ciphers:
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CCM (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CCM_8 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_256_CBC_SHA (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_256_CCM (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_256_CCM_8 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 (dh 2048) - A
|       TLS_DHE_RSA_WITH_ARIA_128_GCM_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_ARIA_256_GCM_SHA384 (dh 2048) - A
|       TLS_DHE_RSA_WITH_CAMELLIA_128_CBC_SHA (dh 2048) - A
|       TLS_DHE_RSA_WITH_CAMELLIA_128_CBC_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA (dh 2048) - A
|       TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (dh 2048) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_ARIA_128_GCM_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_ARIA_256_GCM_SHA384 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_CAMELLIA_128_CBC_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_CAMELLIA_256_CBC_SHA384 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256 (secp256r1) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CCM (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CCM_8 (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CCM (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CCM_8 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|       TLS_RSA_WITH_ARIA_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_ARIA_256_GCM_SHA384 (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_128_CBC_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_256_CBC_SHA256 (rsa 2048) - A
|     compressors:
|       NULL
|     cipher preference: client
|_  least strength: A

Nmap done: 1 IP address (1 host up) scanned in 33.33 seconds
```

mnr73 commented:
> Great work. Do you happen to have any automated commands that might test the feature? ie:
> 
> 1. I request a ECDSA cert and run a console command to check it given I only have the HTTPS URL
> 2. I request a RSA cert and run a console command to check it given I only have the HTTPS URL

with this command we can see what is the SSL key type. rsa or ecdsa

`nmap --script ssl-cert -p 443 example.com`

jc21 commented:
Nice. I'll look at adding this in the next release, after the one I'm about to do today.

jc21 commented on frontend/js/app/nginx/proxy/form.ejs:
```diff
@@ -105,6 +114,15 @@
+                        <div class="col-sm-12 col-md-12">
+                            <div class="form-group">
+                                <label class="form-label"><%- i18n('all-hosts', 'ssl-key-type') %></label>
+                                <select name="ssl_key_type" class="form-control custom-select">
```
Since this setting only applies to Certificates, it should be a new column on the Certificates table. And then there should be some method of changing it's type from that list. There's currently no "edit certificate" dialog so that's going to have to be a thing.

You've got to remember, most users won't know or care about this new certificate SSL type so we need to give advanced users the option without hindering or overwhelming simpler users.

mnr73 replied:
fix this. just show select box when user want create new ssl.

jc21 commented on backend/migrations/20241209062244_ssl_key_type.js:
```diff
@@ -0,0 +1,39 @@
+const migrate_name = 'identifier_for_migrate';
```
And you can combine your multiple migrations into a single file

mnr73 replied:
fix this, i add migration name and combine them

mnr73 commented on docker/start-container:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
if generate this file in build time. it's be same for all user that use this and i think this is a security problem.

jc21 replied:
ok maybe, but why would it be different for all users when they are all using the same docker image?

mnr73 replied:
The DH parameter file is used for secure key exchange, and having the same file for all users can compromise security. It’s recommended to generate a unique file per instance to ensure the security of each user’s connection.

mnr73 commented on backend/internal/host.js:
```diff
@@ -228,8 +228,32 @@ const internalHost = {
-	}
+	},
 
+	/**
```
The problem is that when I try to connect to the server with an IoT device, the connection fails. After some research, I found this command:

openssl s_client -connect <SERVER IP>:443

However, this command returns no response.

When I add a default server to one of the Nginx host configurations, everything works correctly. The above command returns a response, and the IoT device can connect to all the hosts configured in Nginx Proxy Manager.

so i add this feature and its work without any problem.

jc21 replied:
Right. Well it was implemented by another contributor a long time ago, that the default HTTPS host returns a bad cipher/ssl cert or something like that. There was a very good reason for that at the time.

The default-site config doesn't apply to HTTPS though, since any certificate assigned to that would always be invalid for a catch-all domain.

Is there no other way you can fetch the ciphers?

mnr73 replied:
I understand that the certificate for a default server would always be invalid. However, I haven't found any other solution. Even when I manually configured Nginx (before switching to Nginx Proxy Manager), I spent a week troubleshooting this issue. Without setting a default server in one of the configurations, IoT devices simply cannot connect.

I believe this issue might be related to how SNI is handled.

mnr73 commented on backend/internal/nginx.js:
```diff
@@ -185,7 +185,7 @@ const internalNginx = {
 		let host             = JSON.parse(JSON.stringify(host_row));
 		const nice_host_type = internalNginx.getFileFriendlyHostType(host_type);
-
+		
```
i remove this.

mnr73 commented on backend/schema/components/certificate-object.json:
```diff
@@ -41,6 +41,15 @@
+			"enum": ["ecdsa", "rsa"],
+			"description": "Type of SSL key (either ecdsa or rsa)"
+		},
+		"default_server": {
```
mention about this topic above.

jc21 replied:
Even so, it's a setting that applies to a HOST not a CERTIFCATE object.

mnr73 replied:
Yes. You'r right. i can move it to host instead certificate. I'll do this tomorrow.

mnr73 replied:
default server removed from certificate object

mnr73 commented:
I think there is a problem in auto test. i test in local every thing was Ok. also revert all commit but it fail again.

localhost14 commented:
When will this be implemented?

cyberdust2k commented:
Is this still in the scope for a future release? I've been relying on this branch for having RSA on my IOT endpoint for a while now. Anything needing to be done here?
